src/services/spr.ts

```diff
@@ -0,0 +1,263 @@
+import express from 'express';
+import subdomain from 'express-subdomain';
+import Dicer from 'dicer';
+import { getDuplicateCECData, getRandomCECData } from '@/database';
+import { getNEXDataByPID, getFriends } from '@/util';
+import { CECData } from '@/models/cec-data';
+import { SendMode, SPRSlot } from '@/types/common/spr-slot';
+
+const spr = express.Router();
+
+function multipartParser(request: express.Request, response: express.Response, next: express.NextFunction): void {
+	const RE_BOUNDARY = /^multipart\/.+?(?:; boundary=(?:(?:"(.+)")|(?:([^\s]+))))$/i;
+	const RE_FILE_NAME = /name="(.*)"/;
+
+	const contentType = request.header('content-type');
+
+	if (!contentType) {
+		return next();
+	}
+
+	const boundary = RE_BOUNDARY.exec(contentType);
+
+	if (!boundary) {
+		return next();
+	}
+
+	const dicer = new Dicer({ boundary: boundary[1] || boundary[2] });
+	const files: Record<string, Buffer> = {};
+
+	dicer.on('part', (part: Dicer.PartStream) => {
+		let fileBuffer = Buffer.alloc(0);
+		let fileName = '';
+
+		part.on('header', header => {
+			const contentDisposition = header['content-disposition' as keyof object];
+			const regexResult = RE_FILE_NAME.exec(contentDisposition);
+
+			if (regexResult) {
+				fileName = regexResult[1];
+			}
+		});
+
+		part.on('data', (data: Buffer | string) => {
+			if (data instanceof String) {
+				data = Buffer.from(data);
+			}
+
+			fileBuffer = Buffer.concat([fileBuffer, data as Buffer]);
+		});
+
+		part.on('end', () => {
+			files[fileName] = fileBuffer;
+		});
+	});
+
+	dicer.on('finish', function () {
+		request.files = files;
+		return next();
+	});
+
+	request.pipe(dicer);
+}
+
+spr.post('/relay/0', multipartParser, async (request, response) => {
+	if (!request.files) {
+		response.sendStatus(400);
+		return;
+	}
+
+	if (!request.pid) {
+		response.sendStatus(400);
+		return;
+	}
+
+	// * Check that the account is a 3DS and isn't banned
+	const accountData = await getNEXDataByPID(request.pid);
+	if (!accountData) {
+		response.sendStatus(400);
+		return;
+	}
+
+	if (!accountData.friendCode || accountData.accessLevel < 0) {
+		response.sendStatus(400);
+		return;
+	}
+
+	const sprMetadataBuffer: Buffer | undefined = request.files['spr-meta'];
+	if (typeof sprMetadataBuffer === 'undefined') {
+		response.sendStatus(400);
+		return;
+	}
+
+	let sprSlots: SPRSlot[] = [];
+
+	// * Check spr-meta metadata headers
+	const sprMetadata = sprMetadataBuffer.toString();
+	const metadataHeaders = sprMetadata.split('\r\n'); // * Split header lines
+	if (metadataHeaders.length < 1) {
+		response.sendStatus(400);
+		return;
+	}
+
+	for (let i = 0; i < metadataHeaders.length; i++) {
+		const metadataHeader = metadataHeaders[i];
+		const field = metadataHeader.split(': '); // * Split header and value
+		if (field.length !== 2) {
```

Comment:
Also possible refactoring opportunity: JavaScript supports array destructuring. So rather than `field` being assigned and then indexed, you could do `[name, value] = split`

```suggestion
		const field = metadataHeader.split(': '); // * Split header and value

		if (field.length !== 2) {
```

Reply:
It does not. But you can easily just check for the existence of the variables after the 1st. The 1st variable will always be guaranteed to have a value, but all others do not have this guarantee.

For example the following:

```js
const [v1, v2] = 'test'.split(':');

console.log(v1, v2);
```

This will log `test undefined`. However you can set a default value as well:

```
const [v1, v2 = 'default'] = 'test'.split(':');

console.log(v1, v2);
```

Which now prints `test default`

So you can use destructuring and validate that the second field actually has a value or not before continuing